File: Python/1.py
"""Trie-based Autocomplete with Top-5 Suggestions by Frequency"""

from collections import defaultdict

class TrieNode:
    def __init__(self):
        self.children = {}
        self.freq = 0
        self.top5 = []  # [(freq, word), ...]

def insert(root, word):
    node = root
    for ch in word:
        if ch not in node.children:
            node.children[ch] = TrieNode()
        node = node.children[ch]
    node.freq += 1

def get_all_words(node, prefix):
    """Collect all words under this node with their frequencies."""
    results = []
    if node.freq > 0:
        results.append((node.freq, prefix))
    for ch, child in node.children.items():
        results.extend(get_all_words(child, prefix + ch))
    return results

def build_top5(node, prefix):
    """Pre-compute top-5 most frequent words at every trie node."""
    all_words = get_all_words(node, prefix)
    all_words.sort(reverse=True)
    node.top5 = all_words[:5]
    for ch, child in node.children.items():
        build_top5(child, prefix + ch)

def get_suggestions(root, prefix):
    """Return top-5 suggestions for a given prefix."""
    node = root
    for ch in prefix:
        if ch not in node.children:
            return []
        node = node.children[ch]
    return [word for _, word in node.top5]

if __name__ == "__main__":
    root = TrieNode()
    words = ["apple", "app", "application", "ape", "bat", "ball", "batter"]
    for w in words:
        insert(root, w)

    build_top5(root, "")
    prefix = "app"
    suggestions = get_suggestions(root, prefix)
    print(f"Suggestions for '{prefix}': {' '.join(suggestions)}")

File: C++/1.cpp
#include <bits/stdc++.h>
using namespace std;

struct Trie {
    unordered_map<char, Trie*> children;
    int freq = 0;
    vector<pair<int, string>> top5; 
};
void insert(Trie* root, const string& word) {
    Trie* node = root;
    for (char c : word) {
        if (node->children.find(c) == node->children.end()) {
            node->children[c] = new Trie();
        }
        node = node->children[c];
    }
    node->freq++;
}

vector<pair<int, string>> getAllWords(Trie* node, const string& current) {
    vector<pair<int, string>> res;
    if (node->freq > 0) {
        res.push_back({node->freq, current});
    }
    for (auto& p : node->children) {
        auto sub = getAllWords(p.second, current + p.first);
        res.insert(res.end(), sub.begin(), sub.end());
    }
    return res;
}

void buildTop5(Trie* node, const string& current) {
    vector<pair<int, string>> all = getAllWords(node, current);
    sort(all.rbegin(), all.rend());
    for (int i = 0; i < min(5, (int)all.size()); ++i) {
        node->top5.push_back({all[i].first, all[i].second});
    }
    for (auto& p : node->children) {
        buildTop5(p.second, current + p.first);
    }
}

vector<string> getSuggestions(Trie* root, const string& prefix) {
    Trie* node = root;
    for (char c : prefix) {
        if (node->children.find(c) == node->children.end()) {
            return {};
        }
        node = node->children[c];
    }
    vector<string> res;
    for (auto& p : node->top5) {
        res.push_back(p.second);
    }
    return res;
}

int main() {
    Trie* root = new Trie();
    vector<string> words = {"apple", "app", "application", "ape", "bat", "ball", "batter"};
    for (const auto& word : words) {
        insert(root, word);
    }
    
    buildTop5(root, "");
    string prefix = "app";
    vector<string> suggestions = getSuggestions(root, prefix);
    cout << "Suggestions for '" << prefix << "': ";
    for (const auto& s : suggestions) {
        cout << s << " ";
    }
    cout << endl;
    return 0;
}